ooni/agent/scheduler.py

```diff
@@ -82,18 +83,28 @@ def __init__(self, schedule=None, identifier=None,
         assert self.identifier is not None, "self.identifier must be set"
         assert self.schedule is not None, "self.schedule must be set"
 
+        # XXX: both _last_run_lock and _smear_coef require that there is single
+        # instance of the ScheduledTask of each type identified by `identifier`.
         self._last_run = FilePath(scheduler_directory).child(self.identifier)
         self._last_run_lock = FileSystemlockAndMutex(
             FilePath(scheduler_directory).child(self.identifier + ".lock").path
         )
+        self._smear_coef = random.random()
 
     def cancel(self):
+        if not self._last_run_lock.locked:
+            # _last_run_lock.release() will throw if we try to release it
+            log.err('BUG: cancelling non-locked task {} without holding lock'.format(self.identifier))
+            return
+        # probably, cancelling the task TAKEN the lock is even worse :-)
         self._last_run_lock.release()
```

Comment:
Yes, this logic here is very fragile. You are right to add this extra check. 

This was created within the context of being able to reload the list of decks that are currently enabled (`refresh_deck_list`). We need in there to be able to unschedule decks that have been disabled by a user.

Other issues with this method are:

1. If a task is currently running, it will continue execution and not interrupt while it's running

2. Because of 1. if a task was unscheduled and then rescheduled while another one was running, you will end up with two tasks running at the same time. If you toggle a certain task many times this can lead you having a bunch of tasks running at the same time

There are possibly also other issues that I haven't yet thought of.